components/places/src/storage.rs

```diff
@@ -399,6 +395,70 @@ mod tests {
         }
     }
 
+    #[test]
+    fn test_visit_counts() {
+        let mut conn = PlacesDb::open_in_memory(None).expect("no memory db");
+        let url = Url::parse("https://www.example.com").expect("it's a valid url");
+        let early_time = SystemTime::now() - Duration::new(60, 0);
+        let late_time = SystemTime::now();
+
+        // add 2 local visits - add latest first
+        let rid1 = apply_observation(&mut conn, VisitObservation::new(url.clone())
+                    .with_visit_type(VisitTransition::Link)
+                    .with_at(Some(late_time.into())))
+                    .expect("Should apply visit").expect("should get a rowid");
+
+        let _rid2 = apply_observation(&mut conn, VisitObservation::new(url.clone())
+                    .with_visit_type(VisitTransition::Link)
+                    .with_at(Some(early_time.into())))
+                    .expect("Should apply visit").expect("should get a rowid");
+
+        let mut pi = fetch_page_info(&conn, &url).expect("should not fail").expect("should have the page");
+        assert_eq!(pi.page.visit_count_local, 2);
+        assert_eq!(pi.page.last_visit_date_local, late_time.into());
+        assert_eq!(pi.page.visit_count_remote, 0);
+        assert_eq!(pi.page.last_visit_date_remote.0, 0);
+
+        // 2 remote visits, earliest first.
+        let rid3 = apply_observation(&mut conn, VisitObservation::new(url.clone())
+                    .with_visit_type(VisitTransition::Link)
+                    .with_at(Some(early_time.into()))
+                    .with_is_remote(true))
+                    .expect("Should apply visit").expect("should get a rowid");
+
+        let _rid4 = apply_observation(&mut conn, VisitObservation::new(url.clone())
+                    .with_visit_type(VisitTransition::Link)
+                    .with_at(Some(late_time.into()))
+                    .with_is_remote(true))
+                    .expect("Should apply visit").expect("should get a rowid");
+
+        pi = fetch_page_info(&conn, &url).expect("should not fail").expect("should have the page");
+        assert_eq!(pi.page.visit_count_local, 2);
+        assert_eq!(pi.page.last_visit_date_local, late_time.into());
+        assert_eq!(pi.page.visit_count_remote, 2);
+        assert_eq!(pi.page.last_visit_date_remote, late_time.into());
+
+        // Delete some and make sure things update.
+        // XXX - we should add a trigger to update frecency on delete, but at
+        // this stage we don't "officially" support deletes, so this is TODO.
```

Comment:
Doing it via a trigger is going to be tricky (impossible?) since it's not a SQL function, nor can it be really.

Oh well, we'll... figure that out then.

Reply:
Oh, because we can't access the connection from within the function? We _could_ try to make the frecency update function a trigger instead (in other words, rewrite Desktop's `CalculateFrecencyFunction` in pure SQL), but, um...that might be getting too far into the weeds.